src/js/stores/MesosStateStore.js

```diff
@@ -90,79 +49,51 @@ class MesosStateStore extends GetSetBaseStore {
       listenAlways: true
     });
 
-    this.dispatcherIndex = AppDispatcher.register(payload => {
-      if (payload.source !== ActionTypes.SERVER_ACTION) {
-        return false;
-      }
-
-      var action = payload.action;
-      switch (action.type) {
-        case ActionTypes.REQUEST_MESOS_STATE_SUCCESS:
-          this.processStateSuccess(action.data);
-          break;
-        case ActionTypes.REQUEST_MESOS_STATE_ERROR:
-          this.processStateError(action.xhr);
-          break;
-        case ActionTypes.REQUEST_MESOS_STATE_ONGOING:
-          this.processOngoingRequest();
-          break;
-      }
-
-      return true;
+    METHODS_TO_BIND.forEach(method => {
+      this[method] = this[method].bind(this);
     });
-
-    VisibilityStore.addChangeListener(
-      VISIBILITY_CHANGE,
-      this.onVisibilityStoreChange.bind(this)
-    );
   }
 
   addChangeListener(eventName, callback) {
     this.on(eventName, callback);
 
-    if (this.shouldPoll()) {
-      startPolling();
+    if (this.shouldSubscribe()) {
+      this.subscribe();
     }
   }
 
   removeChangeListener(eventName, callback) {
     this.removeListener(eventName, callback);
 
-    if (!this.shouldPoll()) {
-      stopPolling();
+    if (!this.shouldSubscribe()) {
+      this.unsubscribe();
     }
   }
 
-  onVisibilityStoreChange() {
-    if (!VisibilityStore.isInactive() && this.shouldPoll()) {
-      startPolling();
-
-      return;
-    }
-
-    stopPolling();
+  shouldSubscribe() {
+    return this.listeners(MESOS_STATE_CHANGE).length > 0 && !this.stream;
   }
 
-  shouldPoll() {
-    return this.listeners(MESOS_STATE_CHANGE).length > 0;
+  subscribe() {
+    const mesosStream = container.get(MesosStreamType);
+    const parsers = new Pipeline(Object.values(mesosStreamParsers));
+
+    this.stream = mesosStream
+      .merge(request({ type: "GET_MASTER" }))
```

Comment:
🤔 Wonder if we should add a retry to catch network issues.

Reply:
➡️  Please catch potential `GET_MASTER` errors to make sure we're not failing the entire stream if the request fails.

Reply:
I bet the final `onStreamError` catches them, doesn't it?

Reply:
Did a quick test it should catch errors from merged Observables

Reply:
Sure, but I'd only drop the `GET_MASTER` request if it fails instead of the entire stream.

Reply:
I think this should be sufficient:
```javascript
this.stream = mesosStream.merge(request({ type: "GET_MASTER" }).retry(3))...
```
How do you think?

Reply:
Yep. That should be sufficient.